### PR TITLE
fix(ui-server): dedup duplicate station identity on add (#2026)

### LINKED ISSUE #2026: [BUG] Duplicate station identity (same hashId) makes a broadcast UI command report false success
### Duplicate issue

- [X] I've searched for any related issues and avoided creating a duplicate issue.

### Component

Simulator

### Description

Follow-up to #2018 / PR #2020 (explicitly deferred there).

`hashId` is a deterministic content hash with **no uniqueness check on station add**, so two stations built from identical template identity fields collide on one key. PR #2020's `hashId`-keyed completion `Set` cannot disambiguate them, so a targeted UI command completes `success` on the first worker's reply while the collided twin is silently orphaned.

Root cause (verified at commit `c244e582`):
- Deterministic hash of template identity fields + composed id, no per-instance salt: `getHashId` (`src/charging-station/HelpersId.ts:75-103`); `fixedName` → bare `baseName` (`src/charging-station/HelpersId.ts:57-58`).
- No collision check on add: `Bootstrap.addChargingStation` (`src/charging-station/Bootstrap.ts:196-217`) → `workerEventAdded` (`src/charging-station/Bootstrap.ts:726-729`) → `AbstractUIServer.setChargingStationData` (`src/charging-station/ui-server/AbstractUIServer.ts:336-343`) is last-writer-wins by timestamp, keyed by `hashId`, and silently overwrites.
- Completion cannot represent two workers sharing one id: `outstandingHashIds: Set<string>` (`src/charging-station/ui-server/ui-services/AbstractUIService.ts:73`), `recordBroadcastChannelResponse` (`AbstractUIService.ts:203-216`).

### Version

4.10.1 (follow-up to #2020; verified at commit c244e582)

### Node.js version

>= 22.0.0 (project engines). Source-review follow-up from PR #2020, not a runtime environment repro.

### System

N/A — source-code defect identified by review of PR #2020, not a runtime-environment reproduction.

### Expected result

A targeted broadcast command aggregates one reply per distinct physical station; two stations sharing an identity are either prevented at add time or both accounted for. No command reports `success` while leaving a station unactioned.

### Actual result

The first reply removes `H` from the outstanding set → set empties → the request completes `success`; the twin's reply finds `H` gone → `untracked` → dropped. The operator sees success while one station is orphaned/shadowed.

### Steps to reproduce

1. Provision two stations from one template with identical identity fields (e.g. `fixedName: true`) so the composed `chargingStationId` — and therefore `hashId` — is identical.
2. Both are created; `setChargingStationData` overwrites, so the UI registry shows one.
3. Issue a targeted command (e.g. `startChargingStation`) for that `hashId`: it returns `success` on the first worker's reply; the second worker's outcome is dropped.

### Attachments

Proposed fix (scoped): post-creation identity dedup on `workerEventAdded`, **keeping deterministic hashIds** (they back the `<hashId>.json` persisted config, the registry map key, broadcast routing and stats) — reject (→ `hashIdsFailed`) or return-existing when the `hashId` already maps to a different station identity; distinguish a legitimate restart re-emit (newer timestamp) from a distinct twin via a stable instance discriminator such as `stationInfo.templateIndex`.

Acceptance criteria:
- Collision on add is rejected/coalesced with a typed error, never a silent overwrite.
- A legitimate restart/re-emit with a newer timestamp still updates the registry (no regression).
- `getHashId` output stays byte-identical for a given template + index (existing config files resolve).
- Test: (a) twin rejected; (b) restart accepted; (c) no false `success` afterward.

Out of scope / related: reworking aggregation to identity-based tracking; changing the hash algorithm. Causally linked to #2027 (an orphaned twin is exactly the worker #2027 cannot individually terminate).

### PR DESCRIPTION
**Pull request by jerome-benoit**
## Problem

`hashId` is a deterministic content hash of a template's identity fields + composed `chargingStationId`, with no uniqueness check on add. `getHashId` hashes vendor/model + serial-number prefixes + meterType + the composed id — **not** the template file name — and `fixedName: true` collapses the composed id to the bare `baseName`. So two stations built from identical identity fields collide on one `hashId`. `AbstractUIServer.setChargingStationData` was last-writer-wins by timestamp keyed by `hashId`, so the twin silently overwrote the first station in the UI registry; a targeted broadcast UI command (e.g. `startChargingStation`) then completed `success` on the first worker's reply while the collided twin was orphaned — the operator saw success with one station unactioned.

## Fix (scoped per issue)

Post-creation identity dedup at the registry write choke point, keeping deterministic `hashId`s byte-identical:

- `setChargingStationData` returns a discriminated `'set' | 'stale' | 'collision'` outcome. A write whose `hashId` already maps to a **different physical station identity — a distinct `(templateName, templateIndex)` pair —** is rejected as `collision` without overwriting. The guard sits at the single write choke point, so it covers every worker event (added/started/stopped/updated), replacing last-writer-wins with first-writer-wins.
- The `(templateName, templateIndex)` pair is the correct discriminator: `getHashId` excludes the template file name and the index space is per-`templateName`, so two identity-clone template files (copy+rename) can share both `hashId` and `templateIndex`; comparing the pair rejects them, whereas `templateIndex` alone would silently overwrite.
- A legitimate restart/re-emit keeps the same `(templateName, templateIndex)` and still updates by timestamp (no regression).
- `workerEventAdded` logs the rejected collision at `error` level (with `hashId`, `templateName`, `templateIndex`) instead of silently logging the twin as added.

`getHashId` output is byte-identical, so existing `<hashId>.json` config files, the registry map key, broadcast routing and stats resolve unchanged. No OCPP PDU / message-format change; simulator station-registry lifecycle only.

## Tests

- twin (same `hashId`, different `templateIndex`) rejected and not overwritten;
- identity-clone twin (same `hashId` + `templateIndex`, different `templateName`) rejected;
- restart re-emit (same identity, newer timestamp) still updates; stale older-timestamp re-emit dropped;
- no false broadcast `success` after a collision.

Mutation-verified: reverting the collision guard fails the twin/false-success tests; reverting the `templateName` clause fails only the clone-file test; the restart/stale tests stay green throughout. Full project-script gates pass (format/lint/typecheck/test — 3101/3101, 6 pre-existing skips).

## Scope & known limitations (out of scope, deferred)

- Reworking broadcast aggregation to identity-based tracking, and changing the hash algorithm / adding a per-instance salt, are explicitly out of scope.
- Because two twins share one `hashId`, the collided twin's worker `add` genuinely succeeds and its `hashId` still appears in the `addChargingStations` response `hashIdsSucceeded` even though it is absent from the registry; and a later `deleteChargingStations` on that shared `hashId` evicts the surviving primary. Both are inherent to the shared-`hashId` root cause and require the out-of-scope identity-based keying to resolve. This PR keeps the registry and broadcast completion **consistent** for the one accounted station and surfaces the duplicate at add time.
- A collided twin's later `started`/`stopped`/`updated` events return `collision` and are silently dropped (no re-notify); each orphan is announced exactly once at its guaranteed `added` event, so higher-frequency lifecycle events are not re-logged to avoid noise.

Causally linked to #2027 (orphan-worker termination): an orphaned twin is exactly the worker #2027 cannot individually terminate — this add-time dedup surfaces the duplicate instead of letting it silently shadow its twin.

Closes #2026.

### CODE REVIEW

**Comment by jerome-benoit** [comment]
## Review (multi-reviewer cross-validation)

Reviewed against: design/implementation elegance, algorithmic correctness, OCPP-spec compliance, TS/JS best practices, harmonization, content/terminology, scope adherence. Method: three independent reviewers on the same subject (two reasoning-only, one empirical — reads the code and runs typecheck + the affected test file + throwaway edge-case probes), then cross-validation of their outputs.

**Consensus verdict: approve-with-nits → nits resolved in `f84565a7`.**

### Cross-validated findings

**BLOCKING — none.** All three confirmed the choke-point placement (guarding all 4 worker events, not just `workerEventAdded`) is correct and superior; `getHashId`/persisted config/stats/routing untouched; OCPP wire impact N/A (serial numbers are `0..1` optional/informational, no protocol-level identity uniqueness — confirmed against the OCPP 1.6/2.0.1 specs); mathematics/physical-model N/A.

**Unanimous SHOULD-FIX (resolved):**
1. The initial `templateIndex`-only discriminator false-accepts **cross-template identity clones** (copy+rename a template, both `fixedName`, same `baseName`): same `hashId` **and** same `templateIndex` → silent overwrite, reintroducing the bug. Fixed by discriminating on the full `(templateName, templateIndex)` pair (`getHashId` excludes the file name; the index space is per-`templateName`). Restart re-emit keeps both fields → still `set`, no regression.
2. JSDoc/error-message/test only referenced `templateIndex`. Synced the `SetChargingStationDataOutcome` doc to the pair, added `templateName` to the rejection log (so an equal-index clone collision is diagnosable), and added a **same-`templateIndex` / different-`templateName`** test. Mutation-verified: reverting the `templateName` clause fails only that new test.

**Accepted as out-of-scope (documented in the PR description, not defects):**
- Colliding twin still appears in `hashIdsSucceeded` while absent from the registry; `deleteChargingStations` on the shared `hashId` evicts the surviving primary — both inherent to the shared-`hashId` root cause; resolved only by the out-of-scope identity-based keying (#2027). The fix keeps the registry + completion consistent for the one accounted station.
- Silent drop of a collided twin's later `started`/`stopped`/`updated` (each orphan is announced once at its guaranteed `added` event; higher-frequency events are not re-logged to avoid spam) — 2/3 reviewers rated NICE-TO-HAVE; kept minimal.

**TS/harmonization:** `'collision' | 'set' | 'stale'` matches the repo's `BroadcastChannelResponseOutcome` / `UIServerRequestPrologueResult` string-union convention; no `any`/`!`; the 4-caller boolean→union migration is complete (no stray truthiness caller).

**Comment by jerome-benoit** [comment]
## Re-review (multi-reviewer cross-validation, round 2 — post-fix state `f84565a7`)

Method: three fresh independent reviewers on the same subject (two reasoning `oracle`, one adversarial `oracle` tasked to *break* it across 8 attack surfaces), then cross-validation.

**Unanimous verdict: approve, no BLOCKING defect.**

Independently confirmed by all three:
- The `(templateName, templateIndex)` pair is the correct per-template live-instance identity key.
- **No false-reject**: a live `ChargingStation` fixes `this.index`/`this.templateFile` for its lifetime and `stationInfo` is built once; the persisted-config path keeps identity via `??=`. A template-file rename/config-reload cannot make the same live station re-emit a different `(templateName, templateIndex)` — the earlier "rename → false-reject" concern is **unfounded**.
- **No false-accept**: two concurrently-live distinct stations cannot share `(templateName, templateIndex)` (index is unique per `templateName` among live holders); the delete→index-reuse reincarnation case is guarded by the `>=` timestamp check (a lagging pre-deletion event is older → `stale`).
- Type-safe (`templateIndex` is required + validated `> 0`, so `.toString()` cannot throw; the `eslint-disable` covers only the optional `chargingStationId`); collision-before-timestamp ordering is correct; `setChargingStationData` is synchronous (no new hang); the broadcast test confirms the orphaned twin is not double-counted (outstanding = 1).
- `getHashId`/persisted config/stats/routing untouched; OCPP wire impact N/A; mathematics/physical-model N/A.

Residual items are all NICE-TO-HAVE or already documented (see the PR description's *Scope & known limitations*): the collided twin still appears in `hashIdsSucceeded` while absent from the registry, and delete-by-shared-`hashId` collapses the shared key — both inherent to shared-`hashId` aggregation and fenced OUT of scope (identity-based tracking, #2027). A collided twin's later `started`/`stopped`/`updated` events are dropped silently, bounded by the guaranteed one-time `error` at its `added` event.

No further code changes were made: the remaining nits would either diverge from the repo's existing string-union caller conventions (exhaustiveness `switch`) or add log noise for marginal value, against the minimal-diff principle. The fix stands as the smallest correct change for the fenced scope.

File: src/charging-station/Bootstrap.ts [hunk 1]
```diff
@@ -773,7 +783,7 @@ export class Bootstrap extends EventEmitter implements IBootstrap {
   }
 
   private readonly workerEventStarted = (data: ChargingStationData): void => {
-    if (this.uiServer.setChargingStationData(data.stationInfo.hashId, data)) {
+    if (this.uiServer.setChargingStationData(data.stationInfo.hashId, data) === 'set') {
```
[the remark this answers is not in the document]
Comment (jerome-benoit) [reply src/charging-station/Bootstrap.ts:786]: False positive — the unconditional `++templateStatistics.started` is required here.

`waitChargingStationsStopped()` waits for `numberOfStartedChargingStations` (= Σ `started`) occurrences of the `stopped` message event, and that event is emitted for **every** worker (`Bootstrap.ts` messageHandler `case stopped`). A collided orphan twin is a real running worker that emits both `started` and `stopped`, so `started` must count it — it increments on the orphan's `started` and decrements on its `stopped` (balanced, returns to 0).

Early-returning on `'collision'` would make `started` undercount by the number of orphans while those orphans still emit `stopped`, so the shutdown wait would resolve before all workers have stopped — the suggestion introduces the desync it aims to prevent. `workerEventAdded` can early-return safely only because it touches no counter (`added` is incremented in `addChargingStation`, not in the handler). Keeping as-is.

File: src/charging-station/Bootstrap.ts [hunk 2]
```diff
@@ -789,7 +799,7 @@ export class Bootstrap extends EventEmitter implements IBootstrap {
   }
 
   private readonly workerEventStopped = (data: ChargingStationData): void => {
-    if (this.uiServer.setChargingStationData(data.stationInfo.hashId, data)) {
+    if (this.uiServer.setChargingStationData(data.stationInfo.hashId, data) === 'set') {
```
[the remark this answers is not in the document]
Comment (jerome-benoit) [reply src/charging-station/Bootstrap.ts:802]: Same as the `workerEventStarted` thread: `--templateStatistics.started` must run for a collided orphan because its `stopped` message event still fires and is consumed by `waitChargingStationsStopped()`. Guarding it would leave `numberOfStartedChargingStations` counting fewer `stopped` events than actually occur, so the shutdown wait would resolve prematurely. The unconditional decrement is balanced against the unconditional increment on `started` and is correct as-is.

File: tests/charging-station/ui-server/ui-services/AbstractUIService.test.ts [hunk 3]
```diff
@@ -96,6 +97,32 @@ const emitWorkerResponse = (
...
+ * @param templateIndex - Stable per-instance discriminator within a template.
+ * @param timestamp - Registry merge timestamp in milliseconds.
+ * @returns Charging station data for the described identity.
+ */
+const createStationDataWithIndex = (
+  hashId: string,
+  templateIndex: number,
+  timestamp: number
+): ChargingStationData =>
+  createMockChargingStationData(hashId, {
```
[the remark this answers is not in the document]
Comment (jerome-benoit) [reply tests/charging-station/ui-server/ui-services/AbstractUIService.test.ts:115]: Correct observation on the shallow merge, but no change needed. The wholesale `stationInfo` replacement is intentional — the helper supplies a complete `ChargingStationInfo` with exactly the identity fields the dedup inspects (`templateName`/`templateIndex`, plus `baseName`/model/vendor to satisfy the type). `started` is a top-level `ChargingStationData` field (not under `stationInfo`), so the `...overrides` spread preserves it; nothing these tests rely on is dropped. The JSDoc already documents the helper's contract, so I'll leave the merge shallow rather than add depth the tests don't need.
